### PR TITLE
Keep the expert-parallel sentinel slots out of the router gradient

### LINKED ISSUE #48687: Wrong router gradient with `experts_implementation="batched_mm"` under expert parallelism
### System Info

- transformers: main (c694707483)
- torch 2.9, 2x H100, nccl
- Also reproduces on CPU with gloo.

### Who can help?

@ArthurZucker @IlyasMoutawwakil

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

EP with `experts_implementation="batched_mm"` gives the router the wrong gradient. The loss is correct, only the gradient is wrong, so nothing in a training run looks off.

```python
# torchrun --nproc_per_node 2 repro.py
import os, tempfile, torch, torch.distributed as dist
from transformers import OlmoeConfig, OlmoeForCausalLM, set_seed
from transformers.distributed import DistributedConfig

dist.init_process_group("nccl")
rank, world = dist.get_rank(), dist.get_world_size()
torch.cuda.set_device(rank)
device = f"cuda:{rank}"

config = OlmoeConfig(vocab_size=99, hidden_size=32, intermediate_size=16, num_hidden_layers=2,
                     num_attention_heads=4, num_experts=8, num_experts_per_tok=2, norm_topk_prob=False)

path = os.path.join(tempfile.gettempdir(), "ep-router-grad")
if rank == 0:
    set_seed(0)
    OlmoeForCausalLM(config).save_pretrained(path, save_original_format=True)
dist.barrier()

model_ep = OlmoeForCausalLM.from_pretrained(
    path, distributed_config=DistributedConfig(tp_size=world, enable_expert_parallel=True))
model_ref = OlmoeForCausalLM.from_pretrained(path).to(device)
for m in (model_ep, model_ref):
    m.train()
    m.set_experts_implementation("batched_mm")   # grouped_mm is fine

torch.manual_seed(0)
input_ids = torch.randint(0, config.vocab_size, (2, 64), device=device)
labels = torch.randint(0, config.vocab_size, (2, 64), device=device)

model_ref(input_ids, labels=labels).loss.backward()
model_ep(input_ids, labels=labels).loss.backward()

if rank == 0:
    ref = dict(model_ref.named_parameters())
    g = model_ep.get_parameter("model.layers.0.mlp.gate.weight").grad
    g = g.full_tensor() if hasattr(g, "full_tensor") else g
    r = ref["model.layers.0.mlp.gate.weight"].grad
    print(f"router grad norm: reference {r.norm():.6e}  EP {g.norm().to(r.device):.6e}")
```

### Outcome

```
router grad norm: reference 2.051176e-04  EP 3.139034e-04
```

The losses match exactly. Only `mlp.gate.weight` is wrong, on every layer, and it is ~50% too large. Every other parameter agrees to 1e-5.

### Expected behavior

The router gradient should match the non-EP reference, as it does with `grouped_mm`.

### Cause

`batched_mm_experts_forward` clamps the EP sentinel slots into a valid expert and relies on their routing weight being zero:

https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/moe.py#L128-L136

```python
expert_ids = expert_ids.clamp(0, self.num_experts - 1)
...
weighted_out = proj_out * sample_weights.unsqueeze(-1)
```

That is right in the forward: the weight is zero so the slot contributes nothing. It is wrong in the backward!, because the gradient of that product with respect to `sample_weights` is `proj_out`, and `proj_out` at a sentinel slot is not zero, it is expert 0 run on that token. So every slot the rank never routed still sends gradient into `top_k_weights`, and from there into the router.

`grouped_mm_experts_forward` in the same file already handles this, masking `proj_out` at both matmuls before the weighted multiply:

```python
sentinel_mask = (expert_ids_g >= self.num_experts).unsqueeze(-1)
...
proj_out = proj_out.masked_fill(sentinel_mask, 0.0)
```

`batched_mm` is just missing it.

Note the EP tests cannot catch this today: `_test_ep_backward_impl` only compares the loss, which is unaffected. #48518 changes it to compare per-parameter gradients, and that is what surfaced this. PR coming

### PR DESCRIPTION
**Pull request by qgallouedec**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48689) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48689)

Fixes #48687: under EP the sentinel slots keep a non-zero expert output, and the zero routing weight hides that in the forward but not in the backward, so the router collects gradient for slots the rank never routed.

Two things differ from what I suggested in the issue.

The mask goes on the routing weights, not on `proj_out`. Both work, I checked, but the weights are `(S,)` against `(S, hidden_dim)` and this is a hot path. `grouped_mm_experts_forward` masks `proj_out` for a different reason, its kernel leaves the sentinel tail rows uninitialised so it needs the forward too.

`tests/kernels/test_moe_experts_forward.py` covers it without two processes, on the experts forward directly: the sentinel slots must get no routing-weight gradient while the routed slots still get one. It fails on main and passes here. A second test pins that the fix changes no output, by pointing the sentinel slots at a different expert and checking the output is unchanged.

Measured with the repro from the issue, tiny Olmoe on 2 GPUs, EP against a non-EP reference of the same checkpoint. `mlp.gate.weight` gradient norm, layer 0:

| | reference | EP |
|---|---|---|
| before | 2.240401e-04 | 2.846920e-04 |
| after | 2.240401e-04 | 2.240401e-04 |

`eager` and `grouped_mm` already matched the reference, so `batched_mm` was the only one affected. The losses are identical throughout, before and after.

`tests/kernels` is 61 passed, and the EP forward/backward tests on olmoe and deepseek_v3 go from 1 failed to 9 passed once they compare gradients rather than only the loss.

One thing I could not explain (@IlyasMoutawwakil maybe?): `EpRouterParallel` already does `router_scores.masked_fill(non_local_mask, 0.0)`, whose backward zeroes the gradient at exactly these slots, so I would have expected the leak to stop there. It does not, and masking inside the experts forward is what removes it. If that means the real bug is in the hook and this is the wrong layer, say so.

### CODE REVIEW

File: tests/kernels/test_moe_experts_forward.py
Comment (vasqu) [inline tests/kernels/test_moe_experts_forward.py:1]: I'd rather we have a test specification within https://github.com/huggingface/transformers/blob/c6947074832360a35cb8d26d1255b2b517fa9097/tests/test_tensor_parallel_mixin.py#L675-L677

Is there a way to catch it there? Are the bounds too big

Comment (qgallouedec) [reply tests/kernels/test_moe_experts_forward.py:1]: Not with what that test compares today. `_test_ep_backward_impl` asserts on the loss, and the loss is bit-identical here. A gradient that is wrong only on the router leaves the forward untouched by construction, which is why this sat unnoticed.

It becomes catchable there as soon as the test compares parameter gradients rather than the loss. #48518 does that (and that is how this surfaced)

Happy to move that gradient comparison into this PR so the spec lands with the fix, and drop it from #48518. Tell me which you prefer.

I'd keep this unit test either way though.

Comment (qgallouedec) [reply tests/kernels/test_moe_experts_forward.py:1]: check https://github.com/huggingface/transformers/pull/48518/changes/34e63c9d8a3dc0a376125938d5e2630c1d6495ba from #48518

Comment (vasqu) [reply tests/kernels/test_moe_experts_forward.py:1]: Ah that makes sense, I thought we would compare gradients and not only the loss :sweat: maybe we have the same loop hole in the tp backward test?

Comment (vasqu) [reply tests/kernels/test_moe_experts_forward.py:1]: No preference until we figure the other PR out ig, I was just confused why it wasnt caught by my addition but that makes sense

Comment (qgallouedec) [reply tests/kernels/test_moe_experts_forward.py:1]: No, the TP one already compares gradients per parameter on top of the loss, so it would catch this. Only the EP test is loss-only

Comment (vasqu) [inline tests/kernels/test_moe_experts_forward.py:1]: Do we really need to keep this file tho? I'd rather prefer to have this discovered through the existing ep tests (but with the additional grad checks)

Would just be nice to check in the order of merging to make sure we adjust the ep bwd test

File: src/transformers/integrations/moe.py [hunk 1]
```diff
@@ -125,11 +125,13 @@ def batched_mm_experts_forward(
     sample_weights = top_k_weights.reshape(-1)  # (S,)
     expert_ids = top_k_index.reshape(-1)  # (S,)
 
+    sentinel_mask = None
     if self._is_expert_parallel:
         # Clamp EP sentinels so `gate_up_proj[expert_ids]` stays in-bounds. Routing weights are already
         # zero at sentinel slots (RouterParallel masks them at dispatch), so the weighted mul drops
         # those contributions — we pay the wasted GEMM compute because batched_mm has no offset to skip.
         # Out-of-place to avoid mutating the caller's routing tensor (a contiguous `reshape(-1)` aliases it).
+        sentinel_mask = expert_ids >= self.num_experts
```
Comment (vasqu) [inline src/transformers/integrations/moe.py]: let's move it out of the comment above it pls

File: src/transformers/integrations/moe.py [hunk 2]
```diff
@@ -163,6 +165,9 @@ def batched_mm_experts_forward(
     )  # (S, hidden_dim)
 
     # Apply routing weights
+    if sentinel_mask is not None:
+        # Zero weights already drop these from the output; the mask keeps them out of the router gradient.
+        sample_weights = sample_weights.masked_fill(sentinel_mask, 0.0)
```
Comment (vasqu) [inline src/transformers/integrations/moe.py:170]: I did check the router as well but there is nothing that indicates that the leak comes from there, esp since the fwd is correct. Kind of implies that autograd read the weights influence here without this

tried some fused backward with the mask as well but its still leaking so unless Ilyas sees an obvious flaw I think this is pretty straightforward and correct

**Review by vasqu — APPROVED** [review]

File: src/transformers/integrations/moe.py (hunk 1 above)
Comment (qgallouedec) [reply src/transformers/integrations/moe.py]: https://github.com/huggingface/transformers/pull/48689/commits/8de52618c0d3124a0ccb34639419d15e84829ace

File: tests/kernels/test_moe_experts_forward.py
Comment (qgallouedec) [reply tests/kernels/test_moe_experts_forward.py:1]: Technically no, fwiw I'd keep it, for the merge-order gap you're pointing at: if this lands before #48518, the EP backward test on main is still loss-only, so nothing guards the fix in between.

Genuinely fine with dropping it once #48518 is in, or now, as you wish.

Comment (vasqu) [reply tests/kernels/test_moe_experts_forward.py:1]: Yep we can keep it until the other PR lands no worries

Comment (IlyasMoutawwakil) [inline tests/kernels/test_moe_experts_forward.py:1]: maybe not under kernels/

**Review by IlyasMoutawwakil — APPROVED** [review]
thanks ! tbh i'm not sure why grads would leak here, my guess is the collective ops in the parallel module (pre/post experts impl)

File: tests/integrations/test_moe.py
Comment (qgallouedec) [reply tests/integrations/test_moe.py:1]: Moved to `tests/integrations/test_moe.py` in 6341bd0b2f, next to the module it covers.
